feature: Infer response model from function return AST

g7AzaZLO opened the pull request:
# feature: Infer response model from function return AST

## Summary

This PR introduces a feature that automatically infers Pydantic response models by analyzing the AST (Abstract Syntax Tree) of the endpoint function's return statement. This enables FastAPI to generate full OpenAPI schemas for endpoints that return dictionary literals, without requiring an explicit `response_model` or complex type annotations.

## Changes

- **`fastapi/utils.py`**: Added `infer_response_model_from_ast` function.
  - Implemented recursive type inference for nested dictionaries and lists.
  - Added support for basic types (`int`, `str`, `bool`, `float`) from literals and type hints.
  - Added support for simple binary operations (e.g., `10 + 5`) and comparisons.
  - **Fix**: Implemented custom AST traversal to strictly ignore return statements inside nested functions, async functions, or classes.
  - **Fix**: Added graceful fallback (returns `None`) when dictionary keys are not strings (e.g. integers), preventing runtime crashes.

- **`fastapi/routing.py`**: Updated `APIRoute` initialization.
  - Falls back to AST inference if `response_model` is not provided or is not a valid Pydantic model.
  - **Fix**: Added a check to ensure `response_model` is not explicitly `None` (e.g. when returning `Response` or `JSONResponse`) before attempting inference.

## Tests

Added `tests/test_ast_inference.py` with comprehensive coverage:
- **Basic Inference**: Simple dictionary returns with primitive types.
- **Nested Structures**: Lists of objects and deep dictionary nesting.
- **Mixed Types**: Lists with different types fall back to `List[Any]`.
- **Expressions**: Inference for calculated values (`1 + 1`).
- **Edge Cases**:
  - Explicit `JSONResponse` returns (inference disabled).
  - Nested functions (inner returns ignored).
  - Invalid dictionary keys (int keys) -> graceful degradation (no schema).
  - Empty structures (`[]`, `{}`).
  - Local variable assignment before return.
  - "Database" simulation: direct return of external function call (no schema) vs. dict construction from external data (partial schema).

## Motivation

This feature significantly reduces boilerplate for simple "RPC-style" endpoints or quick prototypes where the return structure is static and obvious from the code. It improves the developer experience by providing automatic documentation "for free" while maintaining the robustness of the framework.

## Example

### Old version
```Python
from pydantic import BaseModel

# 1. Boilerplate: We must explicitly define Pydantic models for nested structures.
# This creates a lot of extra code for simple, one-off responses.
class CustomerInfo(BaseModel):
    name: str
    vip_status: bool
    # We have to be specific about types here or use extra models
    preferences: Dict[str, Union[bool, str]]

class Item(BaseModel):
    item_id: int
    name: str
    price: float
    in_stock: bool

# 2. Boilerplate: We must define the main response model matching the dict structure.
class OrderResponse(BaseModel):
    order_id: str
    status: str
    total_amount: float
    tags: List[str]
    customer_info: CustomerInfo
    items: List[Item]
    metadata: Optional[Dict[str, Any]] = None

# 3. Explicit Binding: We must explicitly pass the class to `response_model`.
@app.get("/orders/{order_id}", response_model=OrderResponse)
async def get_order_details(order_id: str):
    # The data construction remains the same, but we had to write 
    # ~20 lines of class definitions above just to validate this dict.
    order_data = {
        "order_id": order_id,
        "status": "processing",
        "total_amount": 150.50,
        "tags": ["urgent", "new_customer"],
        "customer_info": {
            "name": "John Doe",
            "vip_status": False,
            "preferences": {"notifications": True, "theme": "dark"},
        },
        "items": [
            {
                "item_id": 1,
                "name": "Laptop Stand",
                "price": 45.00,
                "in_stock": True,
            },
        ],
        "metadata": None,
    }
    return order_data
```

### New version
Endpoint:
```Python
@app.get("/orders/{order_id}")
async def get_order_details(order_id: str) -> Dict[str, Any]:
    order_data: Dict[str, Any] = {
        "order_id": order_id,
        "status": "processing",
        "total_amount": 150.50,
        "tags": ["urgent", "new_customer"],
        "customer_info": {
            "name": "John Doe",
            "vip_status": False,
            "preferences": {"notifications": True, "theme": "dark"},
        },
        "items": [
            {
                "item_id": 1,
                "name": "Laptop Stand",
                "price": 45.00,
                "in_stock": True,
            },
        ],
        "metadata": None,
    }
    return order_data
 ```
 Response in swagger:
<img width="882" height="557" alt="image" src="https://github.com/user-attachments/assets/6344fa88-8c19-4dc3-a6ca-63250cb16dd8" />

## Key Differences & Benefits

This feature allows FastAPI to automatically infer the response schema from the endpoint's return statement AST, eliminating the need for explicit Pydantic models for simple use cases.

#### 1. Reduction of Boilerplate
*   **Standard Approach:** Developers must define separate `BaseModel` classes for every response, including nested objects (e.g., `CustomerInfo`, `Item`).
*   **With AST Inference:** The schema is generated directly from the dictionary literal in the code. No extra classes are required for single-use schemas.

#### 2. Improved Developer Experience (DX)
*   **Standard Approach:** Prototyping is slower because modifying the response structure requires updating both the dictionary construction logic and the separate class definition.
*   **With AST Inference:** Developers can write idiomatic Python code (returning a `dict`). The API documentation (OpenAPI) and schema validation update automatically to reflect the code structure.

#### 3. Code Locality
*   **Standard Approach:** The schema definition is decoupled from the data generation, often located in different parts of the file or different files entirely.
*   **With AST Inference:** The "source of truth" regarding the data structure resides directly within the `return` statement of the endpoint.

#### 4. Automatic Type Safety for Simple Code
*   Projects using plain dictionaries (typed as `Dict[str, Any]`) instantly gain the benefits of automatic OpenAPI documentation and schema validation without needing a massive refactor to convert everything to Pydantic models.

g7AzaZLO committed:
Implement response model inference from endpoint function source code

- Added "infer_response_model_from_ast" function to analyze endpoint functions and infer Pydantic models from returned dictionary literals or variable assignments.

- Updated "APIRoute" to utilize the new inference method when the specified response model is not a subclass of "BaseModel"

g7AzaZLO committed:
Enhance response model inference and improve error handling

- Updated the response model inference logic in "APIRoute" to check for "None" before validating against "BaseModel".
- Refined the "infer_response_model_from_ast" function to handle nested function definitions and invalid dictionary keys, ensuring robust schema generation.
- Added new test cases for edge scenarios to validate the inference behavior.

g7AzaZLO committed:
Add new database-related endpoints and enhance schema validation in tests

- Introduced two new endpoints: "/db/direct_return" and "/db/dict_construction" to simulate database interactions.
- Added a FakeDB class to mock database behavior for testing purposes.
- Enhanced test cases to validate the OpenAPI schema for the new endpoints, ensuring correct response structure and types.

pre-commit-ci-lite[bot] committed:
🎨 Auto format

g7AzaZLO committed:
Refine response model inference logic in APIRoute

- Updated the response model inference to handle cases where the model is None or not a subclass of BaseModel or a dataclass.
- Enhanced the logic to infer the response model from the endpoint function's source code when necessary, improving schema generation accuracy.

g7AzaZLO committed:
Merge branch 'feat/ast-response-inference' of https://github.com/g7AzaZLO/fastapi into feat/ast-response-inference

g7AzaZLO committed:
Refactor code style and improve readability in utils and test files

- Removed unnecessary blank lines and improved formatting for better readability in `fastapi/utils.py`.
- Consolidated return statements in test cases to a single line for consistency in `tests/test_ast_inference.py`.
- Enhanced the structure of dictionary returns in endpoint functions for clarity.

pre-commit-ci-lite[bot] committed:
🎨 Auto format

g7AzaZLO committed:
Enhance response model inference logic in APIRoute and utils

- Updated the response model inference in APIRoute to check for return annotations before inferring the model from the endpoint's source code.
- Added type ignore comments in utils for better type checking compatibility.
- Specified the type of nodes_to_visit in the infer_response_model_from_ast function for improved clarity.

g7AzaZLO committed:
Enhance response model inference in APIRoute and utils

- Introduced a new helper function `_contains_response` to check for response types in return annotations, improving the inference logic in `APIRoute`.
- Updated the `infer_response_model_from_ast` function to prevent model creation when all fields are of type `Any`, ensuring better type information and avoiding unnecessary overrides.

pre-commit-ci-lite[bot] committed:
🎨 Auto format

g7AzaZLO committed:
Enhance response model inference and add edge case tests

- Updated the `infer_response_model_from_ast` function to use `textwrap.dedent` for cleaner source code handling.
- Added multiple test cases in `tests/test_ast_inference.py` to cover various edge cases for response model inference, including functions with different return types and argument annotations.
- Improved type inference for functions returning lists and nested dictionaries, ensuring better schema generation.

pre-commit-ci-lite[bot] committed:
🎨 Auto format

maxachis commented on fastapi/utils.py:
```diff
@@ -258,3 +261,189 @@ def get_value_or_default(
+        for arg in func_def.args.args:
+            if arg.arg == arg_name and arg.annotation:
+                if isinstance(arg.annotation, ast.Name):
+                    if arg.annotation.id == "int":
```
This case (and quite a few other cases) makes use of the [Arrow Anti Pattern](https://wiki.c2.com/?ArrowAntiPattern), which renders the code quite a bit less readable! In this case, it'd be better to replace these nested conditionals with guard clauses. For example, at `if isinstance(node, ast.Name)`:

```python
if not isinstance(node, ast.Name):
    return Any
```

g7AzaZLO committed:
Refactor type inference logic in utils for improved clarity

- Simplified the `_infer_type_from_ast` function by consolidating conditional checks for argument annotations, enhancing readability and maintainability.
- Updated test cases in `tests/test_ast_inference.py` to utilize parameterization for better organization and coverage of edge cases in response model inference.

g7AzaZLO committed:
Merge branch 'feat/ast-response-inference' of https://github.com/g7AzaZLO/fastapi into feat/ast-response-inference

g7AzaZLO committed:
Refactor import statements in utils for compatibility with Pydantic V2

pre-commit-ci-lite[bot] committed:
🎨 Auto format

g7AzaZLO committed:
Add comprehensive tests for response model inference and coverage

g7AzaZLO committed:
Merge branch 'feat/ast-response-inference' of https://github.com/g7AzaZLO/fastapi into feat/ast-response-inference

pre-commit-ci-lite[bot] committed:
🎨 Auto format

g7AzaZLO committed:
Add error handling test for response model creation

pre-commit-ci-lite[bot] committed:
🎨 Auto format

g7AzaZLO commented on tests/test_ast_inference.py:
```diff
@@ -0,0 +1,441 @@
+
+
+def test_infer_type_from_ast_edge_cases() -> None:
+    """Test edge cases for _infer_type_from_ast function."""
```
The code has been fixed. The edge cases are now split into separate parameterized tests:

test_infer_response_model_returns_none - for cases returning None
test_infer_response_model_success - for successful cases with field validation

The concern is no longer applicable. Thanks for pointing this out - the test structure has been improved.

maxachis reviewed (commented):
Looking on this, and bearing in mind that I'm still familiarizing myself with the innards of FastAPI, I think an issue should be made first and the topic open to broader discussion. The changes are fairly complex, with a number of conditionals. Note how many edge cases you've accounted for, as an example -- it's quite possible there are more!

It's also not clear how it reduces boilerplate as yet, since the example given just shows the new mode, not the old!

g7AzaZLO committed:
Enhance AST inference logging and handle multiple return statements

- Added detailed logging for various failure cases in the `infer_response_model_from_ast` function to aid debugging.
- Modified the logic to collect all return statements, ensuring that inference is skipped if multiple returns are detected.
- Introduced a new test case to verify that functions with multiple return statements correctly return None for inference.

pre-commit-ci-lite[bot] committed:
🎨 Auto format
